### ISSUE TITLE
Consuming pullable sources needs careful handling to avoid stack overflow

### ISSUE THREAD

**Issue by mperktold** [callbag/callbag#39]
A common practice to consume pullable sources is to pull an item on every push.
For an example, see [staltz/callbag-iterate](https://github.com/staltz/callbag-iterate).

That works, and is conceptually very nice and simple.

But in practice, when the source has many items, this leads to a stack overflow:
1. The sink pulls an item by calling `source(1)`.
2. Inside that call, the source pushes an item back by invoking `sink(1, item)`.
3. Inside that call, the sink pulls another item by calling `source(1)`.
4. ...

Thus, every pull and every push adds one frame to the stack, until either the source terminates, or the stack overflows.

But with some care, the sink can avoid stack overflow.

In pull request zebulonj/callbag-pump#1, I fixed this for the `pump` operator by pulling in a loop while an item is pushed back, and restarting the loop on every push if needed.

I choosed `pump` so that the proposed solution can be reused for every pipe of operators, because `pump` effectively converts an arbitrary source to a push-based one, thereby eliminating the mentioned problem for all operators further down the pipe.

For example, the following pipe will lead to a stack overflow:

```javascript
pipe(
  range(0, 10000),
  forEach(x => console.log(x))
);
```

But that can be fixed by inserting `pump` before `forEach`:

```javascript
pipe(
  range(0, 10000),
  pump,
  forEach(x => console.log(x))
);
```

On the other hand, some operators do require input sources to be pullable, but they do not have this problem.
For example, [`toIterable`](https://github.com/staltz/callbag-to-iterable) does not directly consume the input source, but converts it to an `Iterable`, which is then consumed iteratively by the caller.

So to summarize:
Properly consuming pullable sources is harder than one might think at first, but not impossible: It can be done either by using a similar strategy as `pump`, or directly applying `pump` before consuming.

**Comment by staltz** [comment]
Hi! Yes this is an issue that needs to be carefully built and preferably in the source, not in the sink. See the implementation of callbag-from-iter

**Comment by mperktold** [comment]
That makes sense, nice solution! 👍

I think this should be pointed out somewhere officially, both the problem and the solution, so people know how to build and consume pullable sources.

**Comment by franciscotln** [comment]
[Range v2.0.0](https://github.com/franciscotln/callbag-range) has been re-written in a similar fashion as [fromIter](https://github.com/staltz/callbag-from-iter) to address this issue

**Comment by mperktold** [comment]
How about an operator that takes a pullable source and wraps it to apply this solution?

It could also be used as an utility for implementing pullable sources in a simple way and wrapping them internally before returning.

Now I don‘t know what such an operator should be called.

Maybe something abstract like `checkPullable` or `guardStack`?
Or rather something more concrete like `loopingPullable` or `finishPushBeforePull`?

**Comment by staltz** [comment]
@mperktold I would recommend simple synchronous pullables to just be implemented as generator functions (Iterables) so then you pass them to `fromIter` which applies the magic. In other words, fromIter is already that solution you suggested.

**Comment by mperktold** [comment]
Good point, since generator functions are quite easy to write.

Also, this way no missbehaving sources are involved, only a generator function and a proper pullable source.

Still, I think that pullable sources and their properties should be defined more formally in a central place.
Many callbag operators claim to work for listenable sources, pullable sources, or both, so a common definition of these terms would ensure interoperability.

To keep the core specification minimal, I created a repository with a [separate spec for pullable sources](https://github.com/mperktold/callbag-pullable).
It is only a draft, so if you want to modify something, just go for it. 😎

What do you think about having such a specification in general? Should the wiki refer to it?

If this is too much off topic, we can also close this issue and open a new one to discuss about the spec proposal.

**Comment by mperktold** [comment]
> As far as I can see, all sinks could issue pulls "just in case" they are connected to an upstream pullable, and all push-only sources could be expected to simply ignore pulls. Am I wrong?

Yes, sinks can just pull the source when they want to get the next item regardless of whether the source is pullable or listenable.

However, the other way around is much harder: You cannot tell a listenable source when it should emit items and when it shouldn't.

In some cases though you really need a pullable source. For example, take [`to-iterable`](https://github.com/staltz/callbag-to-iterable), which converts a source callbag to an `Iterable`. Since `Iterable` is a pull-based API, the input callbag must be pullable as well.

Therefore, I would say that the best way to have all callbags to flexibly work in pull or push scenarios would be to always implement them as pullable, and keep pulling until completion for achieving push behavior.

Btw, this is actually the idea behind [backpressure in RxJava](https://github.com/ReactiveX/RxJava/wiki/Backpressure).

**Comment by mattflix** [comment]
Thanks @mperktold, but I think you slightly missed my point (or perhaps I am missing yours).

There seem to be some callbag implementations that simply don't work together, but really only for lack of implementation, not for lack of possibility, was my point.

For example, I could create a `forEach` sink that never issues pulls, and thus will only work with sources that actively push data.  But, there seems (to me) to be no good reason for this -- my `forEach` sink could _also_ issue pulls "just in case", and yet handle pulled or pushed data emissions all the same.  In fact, I believe the "standard" `forEach` implementation does just that (i.e., is a "hybrid" implementation that works in either scenario).

So, my question is, why aren't all sinks, sources, and operators implemented in this "hybrid" style?  Is it because some "should not" be implemented as such, or some "cannot", or some simply "are not"?  Based on what I've seen in the implementations I've looked at so far, the answer typically seems to be that they just "are not" (i.e., there is no good reason, other than lack of a "more complete" implementation).

In any case, users of these "non-hybrid" (i.e., pull-only or push-only) implementations are at a bit of a loss to figure out what is going to work and not work, and that seems concerning, and feels like an unnecessary friction in the eco-system.

**Comment by franciscotln** [comment]
@StreetStrider you're right. Thank's for noticing. I'll get that fixed ASAP.

**Comment by mattflix** [comment]
The whole `got1` and `inLoop` riff seems a bit over-complex.

The following approach seems more concise, and is how I've implemented my pullables:

```javascript
let pulls = 0;
sink(0, t => {
  if ((t === 1) && (pulls++ === 1))  {
    do  {
       // push down t=1|2 here
    } while (--pulls > 0);
  } else if (t === 2) {
    pulls = undefined;
  }
})
```

The code says:  "On a pull, start pushing if applicable, and while there are additional pulls, keep pushing, but once we have pushed for every applicable pull, become idle again."

Conveniently, both the initial entry to the loop and exit take the correct action when `pulls` is `undefined` (i.e., does not enter the loop, or exits the loop).

Note that `pulls` should be set to `undefined` whenever the last item is pushed (for finite sources) or the source receives a 2 (neither case shown here).

**Comment by franciscotln** [comment]
> @franciscotln looks like your `got1` is global in callbag-range. I don't think that's intentional.

solved @StreetStrider

**Comment by franciscotln** [comment]
I've proposed this PR to @staltz to simplify the code of `callbag-from-iter`:
[PR#5](https://github.com/staltz/callbag-from-iter/pull/5/files)

BTW @Andarist suggested implementing callbag-range using callbag-from-iter as dependency, which made sense to me and it's done this way in `range v.3.0.0`

**Comment by franciscotln** [comment]
now I think I've got it simpler and working in all cases @Andarist :
```js
const fromIter = iter => (start, sink) => {
  if (start !== 0) return;
  const iterator = typeof Symbol !== 'undefined' && iter[Symbol.iterator]
    ? iter[Symbol.iterator]()
    : iter;
  let value, done = false, disposed = false, looping = false;
  sink(0, t => {
    if (disposed || done) return;
    if (t === 1) {
      while (!looping) {
        looping = true;
        ({ done, value } = iterator.next());
        done ? sink(2) : sink(1, value);
      }
      looping = false;
    }
    if (t === 2) disposed = true;
  });
};
```

**Comment by mattflix** [comment]
No, your while loop is still pushing everything from the iterable whether the sink has pulled it or not.

**Comment by franciscotln** [comment]
Based on what you are saying that it's wrong if it works exactly as the current fromIter? :-)
This is a 1:1 ratio otherwise it wouldn't support lazy pulls, and it does.

**Comment by mattflix** [comment]
I think I may have misinterpreted your while loop.  I see what it does now (it does not send everything at once), but I think the `looping` variable feels like it is named somewhat backwards.

But, more importantly, you still have the issue that multiple pull requests can arrive during a single loop iteration that this code won't properly take into account (since they will all be "merged" into a single state-change of the `looping` variable).  A counter (of the number of pending pulls) would be better (and also leads to a much simpler implementation).

**Comment by franciscotln** [comment]
How can you tell the number of pending pull of a generator function?
```js
function * fibonacci() {
  let fn1 = 0;
	let fn2 = 1;
	let stop = false;
  while (!stop) {  
    let current = fn1;
    fn1 = fn2;
    fn2 = current + fn1;
    stop = yield current;
  }
}
```

**Comment by mattflix** [comment]
The pulls come from the sink, not the generator.  You just need to track the pulls, and "generate" the matching number of pushes.  When they match, you wait for more pulls.  That way, you only push what is pulled (1:1).

**Comment by franciscotln** [comment]
The code snippet that you edited above doesn't emit any value.
If you could share a https://codesandbox.io would be nice to see and test it :-)

**Comment by mperktold** [comment]
Closing this the discussion seems to be finished long ago.

I still feel the need to document this in some official place, but that‘s maybe another issue.
